thetis/adr_model.py

```diff
@@ -0,0 +1,220 @@
+            model_dict['constants'], and the details of each species
+            should be defined under model_dict['species'].
+            A dictionary obtained from a correctly-formatted YAML file using
+            `yaml.safe_load` should be sufficient.
+
+            **Example**
+
+            .. code-block:: python
+
+                {
+                    'constants': {
+                        'D1': 8e-05,
+                        'D2': 4e-05,
+                        'k1': 0.024,
+                        'k2': 0.06
+                    },
+                    'species': {
+                        'a': {
+                            'diffusion': 'D1',
+                            'name': 'Tracer A',
+                            'reaction': '-a*b**2 + k1*(1-a)'
+                        },
+                        'b': {
+                            'diffusion': 'D2',
+                            'name': 'Tracer B',
+                            'reaction': 'a*b^2 - (k1+k2)*b'
+                        }
+                    }
+                }
+
+        :type model_dict: dict
+        :kwarg lambdify_modules: Optional argument to pass as the `modules`
+            parameter of `sympy.lambdify`.
+        :type lambdify_modules: str or dict, optional
+        """
+        self.constants = {
+            "pi": fd.pi,
+            "e": fd.e
+        }
+        self.species = {}
+        self.constant_keys = set(model_dict["constants"].keys())
+        self.species_keys = set(model_dict["species"].keys())
+        self._validate_keys()
+        for k, v in model_dict["constants"].items():
+            self.constants[k] = float(v)
+        for i, (k, v) in enumerate(model_dict["species"].items()):
+            # Parse reaction and diffusion terms
+            reaction_term = parse_expr(
+                str(v["reaction"]),
+                transformations=ADR_Model.transformations)
+            diffusion_term = parse_expr(
+                str(v["diffusion"]),
+                transformations=ADR_Model.transformations)
+            # Check that all symbols in these terms are recognised
+            self._check_all_symbols_recognised(reaction_term)
+            self._check_all_symbols_recognised(diffusion_term)
+            # Substitute values of constants into the expressions
+            # for the reaction and diffusion terms.
+            reaction_term = self._substitute_constants(reaction_term)
+            diffusion_term = self._substitute_constants(diffusion_term)
+            # Obtain a list of arguments to be passed to the
+            # reaction function
+            reaction_args = list(reaction_term.free_symbols)
+            # Store species-specific data.
+            # The reaction expression will now be converted into
+            # a callable function.
+            self.species[k] = {
+                "name": v["name"],
+                "index": i,
+                "diffusion": float(diffusion_term),
```

Comment:
So at the minute we are only supporting constant diffusion terms, which is fine. To support spatially varying diffusivity and coefficients, I guess we would need the functionality to load from file or something. Probably best to just leave for now until someone needs to be able to do it.